LIGHTCURVE(CREATE_LIGHTCURVE, REBIN)

kashish2210 opened the pull request:
THIS PR is a sub-branch of #47. Here I have updated the light_curve function from the previous :)
- Improved performance with vectorized operations
- Enhanced type system for error methods
- Added thread safety and modern Julia features
- Improved error handling and documentation
- Added comprehensive testing suite

taging : @fjebaker and @matteobachetti  for review u may review this [commit ](https://github.com/StingraySoftware/Stingray.jl/pull/51/commits/20ed641e22f3ff353f2d764b6c68c4916e8f6f93) only for this PR

kashish2210 committed:
add lightcurve

kashish2210 committed:
remove println :)

matteobachetti reviewed (changes requested):
Thanks @kashish2210! 
There is something going on with the tests, I see a function being overwritten?
```WARNING: Method definition iterate(Stingray.LightCurve{T} where T) in module Stingray at /home/runner/work/Stingray.jl/Stingray.jl/src/lightcurve.jl:1233 overwritten in module Main at /home/runner/work/Stingray.jl/Stingray.jl/test/test_lightcurve.jl:985.
WARNING: Method definition iterate(Stingray.LightCurve{T} where T, Any) in module Stingray at /home/runner/work/Stingray.jl/Stingray.jl/src/lightcurve.jl:1241 overwritten in module Main at /home/runner/work/Stingray.jl/Stingray.jl/test/test_lightcurve.jl:992.
```
Otherwise, I have another question on the approach chosen for the filtering functions, that you find in its own comment.

kashish2210 commented:
Thanks @matteobachetti , i have seen , i have putten accidentally  base.iterate function in test (it was for array interface later it will be used in plots to make iterative plots)
I will remove it from test_cases;)

kashish2210 committed:
remove overwritten function

fjebaker commented on src/lightcurve.jl:
```diff
@@ -0,0 +1,1243 @@
+"""
+Abstract type for all light curve implementations.
+
+This serves as the base type for all light curve structures, enabling
+polymorphic behavior and type-safe operations across different light curve
+implementations.
```
```suggestion
    abstract type AbstractLightCurve{T} end
```

fjebaker commented on src/lightcurve.jl:
```diff
@@ -0,0 +1,1243 @@
+    "Exposure time for each bin"
+    exposure::Vector{T}
+    "Additional computed properties"
+    properties::Vector{EventProperty}
```
```suggestion
    properties::Vector{EventProperty{T}}
```

fjebaker commented on src/lightcurve.jl:
```diff
@@ -0,0 +1,1243 @@
+    "Time bin centers"
+    timebins::Vector{T}
+    "Time bin edges (length = timebins + 1)"
+    bin_edges::Vector{T}
```
```suggestion
    dt::Union{T,Vector{T}}
```
When uniformly sampled, `dt` can be a single value. If non-uniformly sampled, this should be a vector.

kashish2210 committed:
Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
change lightcurve structure and docs

Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update src/lightcurve.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
update

kashish2210 commented:
Hey @matteobachetti, this is the usage of lightcurve functions :) your reviw required 
## Core Types

### `AbstractLightCurve{T}`
Abstract base type for all light curve implementations.

### `EventProperty{T}`
Stores computed properties per time bin (mean energy, hardness ratios, etc.)
```julia
prop = EventProperty{Float64}(:mean_energy, [1.2, 1.5, 1.8], "keV")
```

### `LightCurveMetadata`
Comprehensive metadata container with observational parameters
```julia
metadata = LightCurveMetadata("NICER", "XTI", "Crab", 58000.0, (1000.0, 5000.0), 1.0, headers, extra)
```

### `LightCurve{T}`
Main structure for binned time series with uncertainties and properties
```julia
lc = create_lightcurve(eventlist, 1.0)  # 1-second bins
```

## Key Functions

### Light Curve Creation
```julia
# Basic usage
lc = create_lightcurve(eventlist, binsize)

# With filtering
lc = create_lightcurve(eventlist, 0.1; 
                      tstart=1000.0, tstop=2000.0,
                      energy_filter=(0.5, 10.0))

# Custom filtering (recommended)
filtered_events = eventlist |>
    ev -> filter_time(t -> t > 1000.0, ev) |>
    ev -> filter_energy(e -> e < 10.0, ev)
lc = create_lightcurve(filtered_events, 1.0)
```

### Error Handling
```julia
# Set Poisson errors (default)
set_errors!(lc)

# Set custom Gaussian errors
set_errors!(lc, [0.5, 1.2, 0.8, 1.1])

# Recalculate errors after modifying counts
lc.counts .+= background_correction
calculate_errors!(lc)
```

### Rebinning
```julia
# Rebin to larger time bins
lc_10s = rebin(lc_1s, 10.0)
```

## Utility Functions

### Time Binning
```julia
edges, centers = create_time_bins(start_time, stop_time, binsize)
counts = bin_events(event_times, bin_edges)
```

### Filtering
```julia
filtered_times, filtered_energies, start_t, stop_t = apply_filters(
    times, energies, tstart, tstop, energy_filter
)
```

### Properties Calculation
```julia
properties = calculate_event_properties(times, energies, bin_edges, bin_centers)
```

## Array Interface
Light curves support standard Julia array operations:
```julia
length(lc)              # Number of bins
lc[1]                   # (time, counts) for first bin
lc[1:5]                 # First 5 bins

# Iteration
for (time, counts) in lc
    println("Time: $time, Counts: $counts")
end
```

## Typical Workflow

1. **Load Event Data**: Start with an `EventList`
2. **Apply Custom Filters**: Use filtering functions on event list
3. **Create Light Curve**: Call `create_lightcurve()` with desired binsize
4. **Set Errors**: Use `set_errors!()` for Poisson or custom errors
5. **Analysis**: Access `.time`, `.counts`, `.count_error` for analysis
6. **Rebinning**: Use `rebin()` if different time resolution needed

## Example Complete Analysis
```julia
# Load and filter events
eventlist = readevents("data.fits")
filtered_events = filter_energy(e -> 0.5 ≤ e ≤ 10.0, eventlist)

# Create light curve
lc = create_lightcurve(filtered_events, 1.0)  # 1s bins

# Set Poisson errors
set_errors!(lc)

# Basic analysis
total_counts = sum(lc.counts)
mean_rate = mean(lc.counts)
peak_time = lc.time[argmax(lc.counts)]

# Rebin for broader view
lc_10s = rebin(lc, 10.0)
```

## Error Methods
- `:poisson`: σ = √N (default, σ = 1 for N = 0)
- `:gaussian`: User-provided custom errors

## Key Features
- Comprehensive metadata preservation
- Flexible filtering system
- Automatic error calculation
- Property computation (mean energy, etc.)
- rebinning
- Full array interface support

kashish2210 commented:
@fjebaker, I have updated this function. Your review is  required 
And I have removed many redundant docs and added efficient docs by using docstrings thanks for the suggestions :)

kashish2210 committed:
adding `set_errors!` ,`calculate_errors!` and updating `create_lightcurve` and `apply_filters`

kashish2210 committed:
exporting functions to stingray module

kashish2210 committed:
updating test case for light curve

kashish2210 committed:
Merge branch 'lightcurve#51' of https://github.com/kashish2210/Stingray.jl into lightcurve#51

kashish2210 committed:
updating lightcurve

kashish2210 committed:
resolving conflicts:!

kashish2210 committed:
Merge branch 'lightcurve#51' of https://github.com/kashish2210/Stingray.jl into lightcurve#51

fjebaker reviewed (approved):
Hi @kashish2210 sorry this has taken me some time to review (travel and conferences are taking up all of my time!). This looks great, thanks for making the changes :)

kashish2210 commented:
No worries, but i have trouble in making this branch history I have maded local changes but previous commit still there

matteobachetti commented:
Let's merge, I will rebase

kashish2210 commented:
> Let's merge, I will rebase
> 

😅😅 Thanks @matteobachetti
